controller/src/util/config_action.rs

```diff
@@ -0,0 +1,429 @@
+            .await
+            .unwrap();
+        }));
+    }
+
+    // Watch for new configs and changes
+    tasks.push(tokio::spawn(async move {
+        watch_for_config_changes(&kube_interface, config_map)
+            .await
+            .unwrap();
+    }));
+
+    futures::future::try_join_all(tasks).await?;
+    info!("do_config_watch - end");
+    Ok(())
+}
+
+/// This watches for Configuration events
+async fn watch_for_config_changes(
+    kube_interface: &impl KubeInterface,
+    config_map: ConfigMap,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    trace!("watch_for_config_changes - start");
+    let resource = Api::<Configuration>::all(kube_interface.get_kube_client());
+    let watcher = watcher(resource, ListParams::default());
+    let mut informer = watcher.boxed();
+    let mut first_event = true;
+    // Currently, this does not handle None except to break the
+    // while.
+    while let Some(event) = informer.try_next().await? {
+        handle_config(kube_interface, event, config_map.clone(), &mut first_event).await?
+    }
+    Ok(())
+}
+async fn get_entry(config_name: &str, config_map: ConfigMap) -> Option<ConfigState> {
+    config_map.lock().await.get(config_name).cloned()
+}
+
+/// This manages the config_map, adding new Configurations and removing old ones.
+/// If the `Configuration::BrokerSpec` was modified, it calls functions to bring down and redeploy
+/// the brokers and services to reflect the updated Configuration.
+async fn handle_config(
+    kube_interface: &impl KubeInterface,
+    event: Event<Configuration>,
+    config_map: ConfigMap,
+    first_event: &mut bool,
+) -> anyhow::Result<()> {
+    trace!("handle_config - something happened to a configuration");
+    match event {
+        Event::Applied(config) => {
+            // Applied events can either be newly added Configurations or modified Configurations.
+            info!(
+                "handle_config - added or modified Configuration {:?}",
+                config.metadata.name.as_ref().unwrap(),
+            );
+            let new_config_state = ConfigState {
+                last_generation: config.metadata.generation,
+                last_configuration_spec: config.spec.clone(),
+            };
+            // Check the Configuration map to see if the Configuration has been updated and
+            // brokers and services need to be redeployed.
+            match get_entry(config.metadata.name.as_ref().unwrap(), config_map.clone()).await {
+                Some(prev_config_state) => match &config.spec.broker_spec {
+                    None => {
+                        config_map.lock().await.insert(
+                            config.metadata.name.as_ref().unwrap().to_string(),
+                            prev_config_state,
+                        );
+                    }
+                    Some(broker) => {
+                        if !should_recreate_config(&config, &prev_config_state) {
```

Comment:
sorry, it has been a while since i've looked at this code ... should_recreate_config is confusing to me.

here we are only acting if it returns false.  it seems like we used to do nothing if it returned false.

the name is also confusing to me.  are we actually recreating a Configuration?  or are we just executing code **as though** the Configuration is deleted and recreated?

Reply:
i agree the name is unhelpful. It works with the agent's context but not the controller. It may be best to rename it something like `broker_changed` or more specifically `only_broker_changed`. This check is done by the agent upon a config change. If true, it does not recreate the configuration rather lets the change stand (no further action). The controller does this check, and if true, it redeploys the brokers to reflect the new config. This is our first step at more explicit configuration modification handling. For this PR, we could leave this out and stick to the old way of deleting and reapplying a configuration, but it means instances will be brought down and redeployed. We need to figure out how to handle this gracefully and this was my first stab at it. it does mean though that if the controller goes down for some reason, a configuration may be applied that does not reflect the system (since the controller was not there to redeploy the brokers)

Reply:
@bfjelds @jiria @romoh on this note, i am thinking that at least for this PR, we should remove support for handling broker changes and stick to our old approach of the agent delete and reapplying modified configurations. We can then spec out and propose separately how we want to handle changes and cooperate between the agent and controller

Reply:
All in favor of splitting into smaller prs.

Another thought on how to handle controller restarts, we could cache the config that was used to create a broker as part of broker annotation. So then controller could check if the broker is based on the latest configuration or not.

Reply:
Good point. That can also done simply via the Job name. I am using the Configuration generation in the Job name. However, someone could feasibly delete a configuration and reapply a new one while the Controller is down and then the Controller would wrongly think it has the right state.